drivers/homeless_summary_report/app/models/homeless_summary_report/report.rb

```diff
@@ -0,0 +1,349 @@
+      filter
+      [
+        build_general_control_section(include_comparison_period: false),
+        build_hoh_control_section,
+        build_coc_control_section(true),
+      ]
+    end
+
+    def report_path_array
+      [
+        :homeless_summary_report,
+        :warehouse_reports,
+        :reports,
+      ]
+    end
+
+    # @return filtered scope
+    def report_scope
+      # Report range
+      scope = report_scope_source
+      scope = filter_for_user_access(scope)
+      scope = filter_for_range(scope)
+      scope = filter_for_cocs(scope)
+      scope = filter_for_head_of_household(scope)
+      scope = filter_for_project_type(scope)
+      scope = filter_for_data_sources(scope)
+      scope = filter_for_organizations(scope)
+      scope = filter_for_projects(scope)
+      scope = filter_for_funders(scope)
+      scope = filter_for_ca_homeless(scope)
+      scope
+    end
+
+    def report_scope_source
+      GrdaWarehouse::ServiceHistoryEnrollment.entry
+    end
+
+    def can_see_client_details?(user)
+      user.can_access_some_version_of_clients?
+    end
+
+    def yn(boolean)
+      boolean ? 'Yes' : 'No'
+    end
+
+    private def create_universe
+      report_clients = {}
+      add_clients(report_clients)
+    end
+
+    private def add_clients(report_clients)
+      spm_reports = run_spm
+
+      # Work through all the SPM report variants, building up the `report_clients` as we go.
+      spm_reports.each do |variant_name, report|
+        self.class.spm_fields.each do |spm_field, cells|
+          cells.each do |cell|
+            spm_clients = answer_clients(report[:report], *cell)
+            spm_clients.each do |spm_client|
+              report_client = report_clients[spm_client[:client_id]] || Client.new
+              report_client[:client_id] = spm_client[:client_id]
+              report_client[:first_name] = spm_client[:first_name]
+              report_client[:last_name] = spm_client[:last_name]
+              report_client[:report_id] = id
+              report_client["spm_#{spm_field}"] = spm_client[spm_field]
+              if self.class.field_measure(spm_field) == 7
+                field_name = "spm_m#{cell.join('_')}".delete('.').downcase.to_sym
+                report_client[field_name] = true
+              end
+              report_client["spm_#{variant_name}"] = report[:report].id
+              report_clients[spm_client[:client_id]] = report_client
+            end
+          end
+        end
+      end
+
+      # With all the fields populated we need to process `exited_from_homeless_system`
+      # TODO: Could we just make this a scope on the client model?
+      report_clients = report_clients.map do |client_id, client|
+        if (
+            client.spm_m7a1_c3 ||
+            client.spm_m7a1_c4 ||
+            client.spm_m7b1_c3
+          ) && !client.spm_m7b2_c3
+          client.exited_from_homeless_system = true
+        end
+        [client_id, client]
+      end.to_h
```

Comment:
This makes an array of all of the clients out of a hash, then turns it back into a hash, which can be fairly RAM/time intensive.  Since these are keyed, let's just use an each and update the value.  (I'll adjust)

Reply:
I see that you used `transform_values` here. I'm pleased to learn about it.

✅